File: iOS/Ortho-iOS/Features/Transactions/Scan/ScanSession.swift
// One capture's lifecycle: source → parse → receipt prefill | interstitial →
// wizard → summary (spec 014, data-model.md §Phase state machine). Owned as
// @State by AddTransactionSheet, so the session — and all capture data —
// dies with the sheet (FR-003).
//
// No transaction ever saves in here: accepts are counted, the sheet performs
// the actual (existing, optimistic) add per row (FR-009).

import SwiftUI
import Observation
import os

@Observable
final class ScanSession {

    enum Source: Equatable {
        case camera, photoLibrary, file
    }

    enum Phase: Equatable {
        case idle
        case parsing
        case receiptPrefilled
        case interstitial
        case reviewing
        case summary
        case failed
    }

    enum Disposition: Equatable {
        case pending
        case added
        case skipped            // user Skip in the wizard
        case leftOutDuplicate   // pre-skipped by the interstitial toggle
        case leftOutPayment     // card-payment row, always pre-skipped (FR-012)
    }

    private(set) var phase: Phase = .idle
    /// Extracted text of the last capture — in-memory only, dies with the
    /// session like everything else here (FR-003). Read by the DEBUG
    /// failure diagnostics so the device itself can show what OCR saw.
    private(set) var lastDocument: ScanDocumentText?
    private(set) var candidates: [ParsedCandidate] = []
    private(set) var dispositions: [UUID: Disposition] = [:]
    private(set) var receiptCandidate: ParsedCandidate?
    /// Wizard queue (pending candidates in document order) + position.
    private(set) var queue: [ParsedCandidate] = []
    private(set) var cursor: Int = 0
    /// Last capture source, so Retake reopens the same one (FR-017).
    var lastSource: Source?

    /// Interstitial toggle, default ON (FR-007). Re-applies pre-skips live.
    var skipDuplicates: Bool = true {
        didSet { if phase == .interstitial { applyPreskips() } }
    }

    // MARK: - Counts (interstitial + summary, FR-010)

    var rowCount: Int { candidates.count }
    var duplicateCount: Int { candidates.filter { $0.duplicateOf != nil && !$0.isPaymentRow }.count }
    var paymentCount: Int { candidates.filter(\.isPaymentRow).count }
    var addedCount: Int { dispositions.values.filter { $0 == .added }.count }
    /// User skips + payment pre-skips read as "skipped" in the summary.
    var skippedCount: Int {
        dispositions.values.filter { $0 == .skipped || $0 == .leftOutPayment }.count
    }
    var leftOutDuplicateCount: Int {
        dispositions.values.filter { $0 == .leftOutDuplicate }.count
    }

    /// Non-nil only while reviewing — callers loop on this (the demo
    /// auto-accept does literally), so it MUST go nil when the wizard ends.
    var currentCandidate: ParsedCandidate? {
        guard phase == .reviewing, queue.indices.contains(cursor) else { return nil }
        return queue[cursor]
    }

    var isLastRow: Bool { cursor >= queue.count - 1 }

    var isWizardActive: Bool { phase == .reviewing }

    // MARK: - Processing

    /// `rescue` is the on-device model's extraction fallback: consulted only
    /// when the deterministic parse yields `.none` on a capture that DID
    /// have text. Nil under -uiDemoScan and in fixtures — determinism holds
    /// where it is contractual, and the spinner stays up while it runs.
    typealias Rescue = @Sendable (ScanDocumentText) async -> ParsedCandidate?

    @MainActor
    func process(images: [UIImage], context: ScanContext, rescue: Rescue? = nil) async {
        await run(context: context, rescue: rescue) { try await ScanTextExtractor.extract(images: images) }
    }

    @MainActor
    func process(pdfAt url: URL, context: ScanContext, rescue: Rescue? = nil) async {
        await run(context: context, rescue: rescue) { try await ScanTextExtractor.extract(pdfAt: url) }
    }

    /// One lifecycle for both sources. The heavy work (OCR / PDF render /
    /// regex parse) runs in a DETACHED task: the app builds with
    /// default-MainActor isolation, so anything less would put Vision's
    /// synchronous recognition on the main thread behind the "Reading…"
    /// spinner (R12).
    @MainActor
    private func run(context: ScanContext,
                     rescue: Rescue?,
                     _ extract: @escaping @Sendable () async throws -> ScanDocumentText) async {
        phase = .parsing
        let (document, parsed) = await Task.detached(priority: .userInitiated) { () -> (ScanDocumentText?, ScanParseResult) in
            do {
                let document = try await extract()
                return (document, ScanParser.parse(document, context: context))
            } catch {
                return (nil, .none)
            }
        }.value
        lastDocument = document
        var result = parsed
        if case .none = result, let rescue, let document, !document.isEmpty,
           let candidate = await rescue(document) {
            result = .receipt(candidate)
        }
        handle(result)
    }

    @MainActor
    private func handle(_ result: ScanParseResult) {
        switch result {
        case .none:
            #if DEBUG
            // The device is the debugger for real-world captures: CI's
            // synthetic fixtures passing ≠ real photos working, and this
            // text never reaches CI logs otherwise.
            if let document = lastDocument, !document.isEmpty {
                Logger(subsystem: Bundle.main.bundleIdentifier ?? "Ortho",
                       category: "scan")
                    .debug("parse failed; extracted: \(document.allLines.map(\.text).joined(separator: " ⏎ "), privacy: .public)")
            }
            #endif
            phase = .failed
        case .receipt(let candidate):
            receiptCandidate = candidate
            phase = .receiptPrefilled
        case .statement(let rows):
            candidates = rows
            dispositions = Dictionary(uniqueKeysWithValues: rows.map { ($0.id, Disposition.pending) })
            applyPreskips()
            phase = .interstitial
        }
    }

    /// Payment rows are always pre-skipped; duplicates follow the toggle.
    /// Nothing here saves or discards permanently — Stop/summary is the only
    /// exit, and pending rows simply never get added.
    private func applyPreskips() {
        for candidate in candidates {
            if candidate.isPaymentRow {
                dispositions[candidate.id] = .leftOutPayment
            } else if candidate.duplicateOf != nil {
                dispositions[candidate.id] = skipDuplicates ? .leftOutDuplicate : .pending
            }
        }
    }

    // MARK: - Wizard flow

    func startReview() {
        queue = candidates.filter { dispositions[$0.id] == .pending }
        cursor = 0
        phase = queue.isEmpty ? .summary : .reviewing
    }

    /// The sheet calls this AFTER performing the row's real optimistic add.
    func acceptCurrent() {
        guard let current = currentCandidate else { return }
        dispositions[current.id] = .added
        advance()
    }

    func skipCurrent() {
        guard let current = currentCandidate else { return }
        dispositions[current.id] = .skipped
        advance()
    }

    /// Always available; rows already added stay (they were real adds).
    func stop() {
        phase = .summary
    }

    private func advance() {
        if isLastRow {
            phase = .summary
        } else {
            cursor += 1
        }
    }

    // MARK: - Lifecycle

    func retake() {
        phase = .idle
    }

    func reset() {
        phase = .idle
        lastDocument = nil
        candidates = []
        dispositions = [:]
        receiptCandidate = nil
        queue = []
        cursor = 0
        lastSource = nil
    }
}

File: web/lib/scan/scanSession.ts
// One capture's lifecycle, ported to a plain TypeScript reducer (spec 021,
// ported from
// iOS/Ortho-iOS/Features/Transactions/Scan/ScanSession.swift).
// Contract: specs/021-capacitor-ios-consolidation/data-model.md
// ("ScanSession (UI state machine)").
//
// Phase state machine: idle -> parsing -> receiptPrefilled | interstitial ->
// reviewing -> summary | failed. Disposition tracking (pending/added/
// skipped/leftOutDuplicate/leftOutPayment) mirrors the Swift enum 1:1 —
// product-approved behavior, not something to redesign.
//
// This module is UI-framework-agnostic beyond matching React's
// `useReducer(reducer, initialState)` shape: `ScanSessionState` is a plain
// immutable value, `ScanSessionAction` is a discriminated union, and
// `scanSessionReducer` is a pure `(state, action) => state` function. The
// `select*` helpers below mirror `ScanSession.swift`'s computed properties
// (`rowCount`, `currentCandidate`, etc.) as plain functions over the state,
// since a reducer has no room for stored computed properties.
//
// No transaction ever saves in here: accepts are counted, the caller (the
// eventual AddTransactionSheet-equivalent, wired in a later task) performs
// the actual optimistic add per row (FR-009) and only then dispatches
// `review/accept`.

import type { ParsedCandidate, ScanDocumentText, ScanParseResult } from './scanModels'

/** Mirrors Swift's `ScanSession.Source`. */
export type ScanSource = 'camera' | 'photoLibrary' | 'file'

/** Mirrors Swift's `ScanSession.Phase`. */
export type ScanSessionPhase =
  | 'idle'
  | 'parsing'
  | 'receiptPrefilled'
  | 'interstitial'
  | 'reviewing'
  | 'summary'
  | 'failed'

/** Mirrors Swift's `ScanSession.Disposition`. */
export type Disposition =
  | 'pending'
  | 'added'
  | 'skipped' // user Skip in the wizard
  | 'leftOutDuplicate' // pre-skipped by the interstitial toggle
  | 'leftOutPayment' // card-payment row, always pre-skipped (FR-012)

/** Keyed by `ParsedCandidate.id` (Swift keys by `UUID`; TS candidates carry
 *  a string `id` already — see scanModels.ts). */
export type DispositionMap = Record<string, Disposition>

/** Why a capture landed on the `failed` phase. `unreadable` is the ordinary
 *  "nothing parseable" outcome (native + web); `unsupportedOnWeb` is the
 *  honest web-only degrade for photo receipts, which need OCR the browser has
 *  no equivalent of (fix/web-scan-fallback) — kept distinct so the UI can show
 *  targeted copy instead of the generic "couldn't read this". */
export type ScanFailureReason = 'unreadable' | 'unsupportedOnWeb'

export interface ScanSessionState {
  phase: ScanSessionPhase
  /** Extracted text of the last capture — in-memory only, dies with the
   *  session like everything else here (FR-003). */
  lastDocument: ScanDocumentText | null
  candidates: ParsedCandidate[]
  dispositions: DispositionMap
  receiptCandidate: ParsedCandidate | null
  /** Wizard queue (pending candidates in document order) + position. */
  queue: ParsedCandidate[]
  cursor: number
  /** Last capture source, so Retake reopens the same one (FR-017). */
  lastSource: ScanSource | null
  /** Interstitial toggle, default ON (FR-007). */
  skipDuplicates: boolean
  /** Only meaningful while `phase === 'failed'`; see `ScanFailureReason`. */
  failureReason: ScanFailureReason
}

export function createScanSessionState(): ScanSessionState {
  return {
    phase: 'idle',
    lastDocument: null,
    candidates: [],
    dispositions: {},
    receiptCandidate: null,
    queue: [],
    cursor: 0,
    lastSource: null,
    skipDuplicates: true,
    failureReason: 'unreadable',
  }
}

export type ScanSessionAction =
  | { type: 'capture/start'; source: ScanSource }
  | { type: 'capture/parsed'; result: ScanParseResult; document?: ScanDocumentText | null }
  | { type: 'capture/unsupported' }
  | { type: 'toggleSkipDuplicates'; value: boolean }
  | { type: 'review/start' }
  | { type: 'review/accept' }
  | { type: 'review/skip' }
  | { type: 'review/stop' }
  | { type: 'retake' }
  | { type: 'reset' }

export function scanSessionReducer(state: ScanSessionState, action: ScanSessionAction): ScanSessionState {
  switch (action.type) {
    case 'capture/start':
      return { ...state, phase: 'parsing', lastSource: action.source }

    case 'capture/parsed':
      return handleParseResult(state, action.result, action.document ?? null)

    // Photo receipts on the web have no OCR (no VisionKit in the browser), so
    // the web camera path degrades honestly instead of silently failing.
    case 'capture/unsupported':
      return { ...state, phase: 'failed', failureReason: 'unsupportedOnWeb', lastSource: 'camera' }

    case 'toggleSkipDuplicates': {
      const skipDuplicates = action.value
      // Re-applies pre-skips live, matching Swift's `didSet { if phase ==
      // .interstitial { applyPreskips() } }`.
      if (state.phase !== 'interstitial') {
        return { ...state, skipDuplicates }
      }
      return {
        ...state,
        skipDuplicates,
        dispositions: applyPreskips(state.candidates, state.dispositions, skipDuplicates),
      }
    }

    case 'review/start': {
      const queue = state.candidates.filter((c) => state.dispositions[c.id] === 'pending')
      return {
        ...state,
        queue,
        cursor: 0,
        phase: queue.length === 0 ? 'summary' : 'reviewing',
      }
    }

    case 'review/accept': {
      const current = selectCurrentCandidate(state)
      if (!current) return state
      return advance({
        ...state,
        dispositions: { ...state.dispositions, [current.id]: 'added' },
      })
    }

    case 'review/skip': {
      const current = selectCurrentCandidate(state)
      if (!current) return state
      return advance({
        ...state,
        dispositions: { ...state.dispositions, [current.id]: 'skipped' },
      })
    }

    // Always available; rows already added stay (they were real adds).
    case 'review/stop':
      return { ...state, phase: 'summary' }

    // Only the phase resets — captured data (candidates/dispositions/queue)
    // is left alone, matching Swift's `retake()` exactly.
    case 'retake':
      return { ...state, phase: 'idle' }

    // Full reset, EXCEPT `skipDuplicates`: Swift's `reset()` never touches
    // it (it is a persisted preference, not capture data).
    case 'reset':
      return { ...createScanSessionState(), skipDuplicates: state.skipDuplicates }

    default:
      return state
  }
}

function handleParseResult(
  state: ScanSessionState,
  result: ScanParseResult,
  document: ScanDocumentText | null
): ScanSessionState {
  switch (result.kind) {
    case 'none':
      return { ...state, lastDocument: document, phase: 'failed', failureReason: 'unreadable' }

    case 'receipt':
      return {
        ...state,
        lastDocument: document,
        receiptCandidate: result.candidate,
        phase: 'receiptPrefilled',
      }

    case 'statement': {
      const dispositions: DispositionMap = {}
      for (const row of result.candidates) dispositions[row.id] = 'pending'
      return {
        ...state,
        lastDocument: document,
        candidates: result.candidates,
        dispositions: applyPreskips(result.candidates, dispositions, state.skipDuplicates),
        phase: 'interstitial',
      }
    }
  }
}

/** Payment rows are always pre-skipped; duplicates follow the toggle.
 *  Nothing here saves or discards permanently — Stop/summary is the only
 *  exit, and pending rows simply never get added. */
function applyPreskips(
  candidates: ParsedCandidate[],
  dispositions: DispositionMap,
  skipDuplicates: boolean
): DispositionMap {
  const next = { ...dispositions }
  for (const candidate of candidates) {
    if (candidate.isPaymentRow) {
      next[candidate.id] = 'leftOutPayment'
    } else if (candidate.duplicateOf != null) {
      next[candidate.id] = skipDuplicates ? 'leftOutDuplicate' : 'pending'
    }
  }
  return next
}

function advance(state: ScanSessionState): ScanSessionState {
  const isLastRow = state.cursor >= state.queue.length - 1
  if (isLastRow) {
    return { ...state, phase: 'summary' }
  }
  return { ...state, cursor: state.cursor + 1 }
}

// MARK: - Selectors (mirror ScanSession.swift's computed properties, FR-010)

export function selectRowCount(state: ScanSessionState): number {
  return state.candidates.length
}

export function selectDuplicateCount(state: ScanSessionState): number {
  return state.candidates.filter((c) => c.duplicateOf != null && !c.isPaymentRow).length
}

export function selectPaymentCount(state: ScanSessionState): number {
  return state.candidates.filter((c) => c.isPaymentRow).length
}

export function selectAddedCount(state: ScanSessionState): number {
  return Object.values(state.dispositions).filter((d) => d === 'added').length
}

/** User skips + payment pre-skips read as "skipped" in the summary. */
export function selectSkippedCount(state: ScanSessionState): number {
  return Object.values(state.dispositions).filter((d) => d === 'skipped' || d === 'leftOutPayment').length
}

export function selectLeftOutDuplicateCount(state: ScanSessionState): number {
  return Object.values(state.dispositions).filter((d) => d === 'leftOutDuplicate').length
}

/** Non-null only while reviewing — callers loop on this (the demo
 *  auto-accept does literally), so it MUST go nil/null when the wizard
 *  ends. */
export function selectCurrentCandidate(state: ScanSessionState): ParsedCandidate | null {
  if (state.phase !== 'reviewing') return null
  return state.queue[state.cursor] ?? null
}

export function selectIsLastRow(state: ScanSessionState): boolean {
  return state.cursor >= state.queue.length - 1
}

export function selectIsWizardActive(state: ScanSessionState): boolean {
  return state.phase === 'reviewing'
}
